jmxtrans-core/src/main/java/com/googlecode/jmxtrans/model/Server.java

```diff
@@ -307,6 +355,40 @@ public String getProtocolProviderPackages() {
 		return protocolProviderPackages;
 	}
 
+	/**
+	 * Factory to create a JMXServiceURL from a pid. Inner class to prevent class
+	 * loader issues when tools.jar isn't present.
+	 */
+	private static class JMXServiceURLFactory {
+
+		public static JMXServiceURL extractJMXServiceURLFromPid(String pid) throws IOException {
```

Comment:
Right now, if tools.jar isn't in the classpath, you get java.lang.NoClassDefFoundError: com/sun/tools/attach/VirtualMachine.

Other frameworks are having pretty much the same behavior. They usually try to guess where is tools.jar to add it to the classpath in their sh. I haven't done it because I was a bit lost in the sh code. So I'll leave that to you.

In case of error, the only other thing we can do is to have a nicer error message. And some documentation (but again, I got lost a bit)